File: allcode/3200-3299/3234numberOfSubstrings.py


from leetcode.allcode.competition.mypackage import *

class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        zero = []
        for i, x in enumerate(s):
            if x == '0':
                zero.append(i)
        if len(zero) == 0:
            return n * (n + 1) // 2
        zero.append(n)
        m = len(zero)
        ans = 0
        idx = 0 if zero[0] > 0 else 1  # l 右侧第一个0在zero中的位置
        for l in range(n):
            if l >= zero[idx]:
                idx += 1
            for r in range(idx, m):
                n_zero = r - idx  # [l, zero[r]) 有多少个0
                if s[l] == '0': n_zero += 1
                n_one = zero[r] - l - n_zero  # [l, zero[r]) 有多少个1
                if n_zero ** 2 > n - l - n_zero:  # [l, zero[r]) 中 0 的个数平方超过 [l, n) 中1的个数的上限
                    break
                if n_zero ** 2 > n_one:
                    continue
                if r > 0:
                    ans += min(zero[r] - max(l, zero[r - 1]), n_one - n_zero ** 2 + 1)
                else:
                    ans += min(zero[r] - l, n_one - n_zero ** 2 + 1)

        return ans


so = Solution()
print(so.numberOfSubstrings("101101"))  # 16
print(so.numberOfSubstrings("1"))  # 16
print(so.numberOfSubstrings("00011"))  # 5

File: c++/src/3000-3999/3234numberOfSubstrings.cpp
#include "lc_pub.h"


class Solution {
    public:
    int numberOfSubstrings(string s) {
        int n=s.size();
        vector<int> p0;
        for (int i=0;i<n;i++) {
            if (s[i]=='0') p0.push_back(i);
        }
        p0.push_back(n);  // 哨兵
        int m=p0.size(), idx=0;
        int ans=0;
        for (int i=0;i<n;i++) {
            // 左端点在 i
            if (idx<m&&p0[idx]<i) idx++;
            if(idx<m) ans+=p0[idx]-i;
            for (int j=idx;j<m-1;j++) {
                // 右端点在 [p0[j], p0[j+1]) 区间内有c0个0
                int c0=j-idx+1;
                int c1=c0*c0; // 至少需要c1个1
                int i1=i+c1+c0-1;  // 右端点至少在i1才能满足，1的个数>=c1
                if (i1>=n) break;
                if (i1>=p0[j+1]) continue;
                ans+=p0[j+1]-max(i1,p0[j]);
            }
        }
        return ans;
    }
    };

    
int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;
    vector<int> nums{5,2,5,4,5};

    Solution so;
    cout<<so.numberOfSubstrings("101101");
    return 0;
}
